File: go/tests/integration_email_test.go
// integration_email_test.go
//
// Verifies the Gmail + MS Graph mail connectors (integration_email.go) over the
// injected FakeCarrierTransport — no real network. Covers Gmail list→per-message
// fetch, header/label/body(base64url) parsing, unread detection, mark-read; and
// MS Graph unread list, search, and mark-read PATCH.

package circleai_test

import (
	"context"
	"encoding/base64"
	"strings"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func b64url(s string) string {
	return strings.TrimRight(base64.URLEncoding.EncodeToString([]byte(s)), "=")
}

func TestGmail_ListUnreadFetchesAndParses(t *testing.T) {
	tr := circleai.NewFakeCarrierTransport()
	// 1) list ids
	tr.EnqueueJSON(200, `{"messages":[{"id":"m1"},{"id":"m2"}]}`)
	// 2) m1 full
	tr.EnqueueJSON(200, `{"id":"m1","labelIds":["INBOX","UNREAD"],"internalDate":"1752226200000",
		"payload":{"headers":[
			{"name":"From","value":"alice@x"},
			{"name":"To","value":"me@y, bob@z"},
			{"name":"Subject","value":"Hi"}],
		"body":{"data":"`+b64url("Hello body")+`"}}}`)
	// 3) m2 full (read, multipart text/plain)
	tr.EnqueueJSON(200, `{"id":"m2","labelIds":["INBOX"],"internalDate":"1752226100000",
		"payload":{"headers":[{"name":"From","value":"carol@x"},{"name":"Subject","value":"Later"}],
		"parts":[{"mimeType":"text/html","body":{"data":"`+b64url("<b>x</b>")+`"}},
		         {"mimeType":"text/plain","body":{"data":"`+b64url("plain part")+`"}}]}}`)

	c := mustGmail(t, tr)
	msgs, err := c.ListUnread(context.Background(), 10)
	if err != nil {
		t.Fatalf("list unread: %v", err)
	}
	if len(msgs) != 2 {
		t.Fatalf("expected 2 messages, got %d", len(msgs))
	}
	m1 := msgs[0]
	if m1.MessageID != "m1" || m1.From != "alice@x" || m1.Subject != "Hi" || !m1.Unread ||
		len(m1.To) != 2 || m1.To[0] != "me@y" || m1.To[1] != "bob@z" || m1.BodyText != "Hello body" {
		t.Fatalf("m1 wrong: %+v", m1)
	}
	if !m1.ReceivedUtc.Equal(time.UnixMilli(1752226200000).UTC()) {
		t.Fatalf("m1 received wrong: %v", m1.ReceivedUtc)
	}
	m2 := msgs[1]
	if m2.Unread || m2.BodyText != "plain part" { // prefers text/plain over html
		t.Fatalf("m2 wrong: %+v", m2)
	}
	// First request carries q=is:unread + Bearer.
	req0 := tr.Requests()[0]
	if !strings.Contains(req0.URL, "messages?q=is%3Aunread") || !strings.HasPrefix(req0.Headers["Authorization"], "Bearer ") {
		t.Fatalf("list request wrong: %s hdr=%v", req0.URL, req0.Headers)
	}
}

func TestGmail_SearchValidationAndSkipOnFetchFailure(t *testing.T) {
	c := mustGmail(t, circleai.NewFakeCarrierTransport())
	if _, err := c.Search(context.Background(), "  ", 5); err == nil {
		t.Fatalf("blank query should error")
	}
	if _, err := c.Search(context.Background(), "x", 0); err == nil {
		t.Fatalf("max<=0 should error")
	}
	// A per-message 404 is skipped (not fatal).
	tr := circleai.NewFakeCarrierTransport()
	tr.EnqueueJSON(200, `{"messages":[{"id":"ok"},{"id":"bad"}]}`)
	tr.EnqueueJSON(200, `{"id":"ok","labelIds":["UNREAD"],"internalDate":"0","payload":{"headers":[]}}`)
	tr.EnqueueStatus(404)
	c2 := mustGmail(t, tr)
	msgs, err := c2.Search(context.Background(), "x", 10)
	if err != nil {
		t.Fatalf("search: %v", err)
	}
	if len(msgs) != 1 || msgs[0].MessageID != "ok" {
		t.Fatalf("expected only the ok message, got %+v", msgs)
	}
}

func TestGmail_MarkRead(t *testing.T) {
	tr := circleai.NewFakeCarrierTransport()
	tr.EnqueueStatus(200)
	c := mustGmail(t, tr)
	if err := c.MarkRead(context.Background(), "m1"); err != nil {
		t.Fatalf("mark read: %v", err)
	}
	req, _ := tr.LastRequest()
	if req.Method != "POST" || !strings.Contains(req.URL, "messages/m1/modify") ||
		!strings.Contains(string(req.Body), `"removeLabelIds":["UNREAD"]`) {
		t.Fatalf("mark-read request wrong: %s %s body=%s", req.Method, req.URL, req.Body)
	}
	if err := c.MarkRead(context.Background(), "  "); err == nil {
		t.Fatalf("blank id should error")
	}
}

func TestGmail_ConfigAndAuth(t *testing.T) {
	c, _ := circleai.NewGmailEmailConnector(circleai.NewFakeCarrierTransport(), circleai.GmailOptions{})
	if c.IsConfigured() {
		t.Fatalf("nil token provider should be unconfigured")
	}
	tr := circleai.NewFakeCarrierTransport()
	c2, _ := circleai.NewGmailEmailConnector(tr, circleai.GmailOptions{AccessTokenProvider: fixedToken("")})
	if _, err := c2.ListUnread(context.Background(), 5); err == nil {
		t.Fatalf("blank token should error")
	}
	if len(tr.Requests()) != 0 {
		t.Fatalf("auth failure issued a request")
	}
}

func TestMsGraphMail_ListUnreadAndSearch(t *testing.T) {
	tr := circleai.NewFakeCarrierTransport()
	tr.EnqueueJSON(200, `{"value":[
		{"id":"e1","subject":"Report","isRead":false,
		 "from":{"emailAddress":{"address":"a@x"}},
		 "toRecipients":[{"emailAddress":{"address":"me@y"}}],
		 "receivedDateTime":"2026-07-11T08:00:00Z",
		 "categories":["Work"],"body":{"content":"full body"}}
	]}`)
	c, _ := circleai.NewMsGraphEmailConnector(tr, circleai.MsGraphEmailOptions{AccessTokenProvider: fixedToken("tok")})
	if c.ProviderID() != "ms-graph-mail" || !c.IsConfigured() {
		t.Fatalf("graph mail id/configured wrong")
	}
	msgs, err := c.ListUnread(context.Background(), 5)
	if err != nil {
		t.Fatalf("list unread: %v", err)
	}
	if len(msgs) != 1 {
		t.Fatalf("expected 1, got %d", len(msgs))
	}
	m := msgs[0]
	if m.MessageID != "e1" || m.From != "a@x" || m.To[0] != "me@y" || m.Subject != "Report" ||
		m.BodyText != "full body" || !m.Unread || m.Labels[0] != "Work" {
		t.Fatalf("graph message wrong: %+v", m)
	}
	if !m.ReceivedUtc.Equal(time.Date(2026, 7, 11, 8, 0, 0, 0, time.UTC)) {
		t.Fatalf("received wrong: %v", m.ReceivedUtc)
	}
	req := tr.Requests()[0]
	if !strings.Contains(req.URL, "mailFolders('Inbox')/messages") || !strings.Contains(req.URL, "isRead+eq+false") {
		t.Fatalf("list url wrong: %s", req.URL)
	}

	// Search path uses $search; bodyPreview fallback when no body.content.
	tr2 := circleai.NewFakeCarrierTransport()
	tr2.EnqueueJSON(200, `{"value":[{"id":"s1","isRead":true,"bodyPreview":"preview only","subject":"S"}]}`)
	c2, _ := circleai.NewMsGraphEmailConnector(tr2, circleai.MsGraphEmailOptions{AccessTokenProvider: fixedToken("tok")})
	res, err := c2.Search(context.Background(), "invoice", 5)
	if err != nil {
		t.Fatalf("search: %v", err)
	}
	if len(res) != 1 || res[0].BodyText != "preview only" || res[0].Unread {
		t.Fatalf("search result wrong: %+v", res)
	}
	if !strings.Contains(tr2.Requests()[0].URL, "me/messages?$search=invoice") {
		t.Fatalf("search url wrong: %s", tr2.Requests()[0].URL)
	}
	if _, err := c2.Search(context.Background(), "  ", 5); err == nil {
		t.Fatalf("blank query should error")
	}
}

func TestMsGraphMail_MarkReadPatch(t *testing.T) {
	tr := circleai.NewFakeCarrierTransport()
	tr.EnqueueStatus(200)
	c, _ := circleai.NewMsGraphEmailConnector(tr, circleai.MsGraphEmailOptions{AccessTokenProvider: fixedToken("tok")})
	if err := c.MarkRead(context.Background(), "e1"); err != nil {
		t.Fatalf("mark read: %v", err)
	}
	req, _ := tr.LastRequest()
	if req.Method != "PATCH" || !strings.Contains(req.URL, "me/messages/e1") ||
		!strings.Contains(string(req.Body), `"isRead":true`) {
		t.Fatalf("patch request wrong: %s %s body=%s", req.Method, req.URL, req.Body)
	}
}

File: typescript/tests/integration_email.test.ts
// integration_email.test.ts
// Verifies the CircleAI.Integration.Email port: Gmail v1 (list-then-get, MIME
// body extraction, base64url), MS Graph mail (unread/search/mark), and the IMAP
// connector over a fake IImapTransport (search → descending slice → fetch/parse).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import type { HttpRequest, HttpResponse, IHttpClient } from "../src/integration/index";
import { DateTimeOffsetMinValue } from "../src/integration/index";
import {
  GmailEmailConnector,
  MsGraphEmailConnector,
  ImapEmailConnector,
  gmailOptions,
  msGraphEmailOptions,
  imapOptions,
  ImapFolderAccess,
  ImapMessageFlags,
  type IImapTransport,
  type ImapMessageSummary,
  type ImapSearchQuery,
} from "../src/integration/email/index";

class FakeHttp implements IHttpClient {
  readonly requests: HttpRequest[] = [];
  constructor(private handler: (r: HttpRequest) => HttpResponse) {}
  send(request: HttpRequest): Promise<HttpResponse> {
    this.requests.push(request);
    return Promise.resolve(this.handler(request));
  }
}

const ok = (body: string): HttpResponse => ({ statusCode: 200, body });
const b64url = (s: string) => Buffer.from(s, "utf8").toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");

describe("GmailEmailConnector", () => {
  const mk = (token: string | null) => new GmailEmailConnector(gmailOptions(async () => token), buildGmailHttp());

  function buildGmailHttp(): FakeHttp {
    return new FakeHttp((r) => {
      if (r.url.includes("messages?q=")) {
        return ok(JSON.stringify({ messages: [{ id: "1" }, { id: "2" }] }));
      }
      if (r.url.includes("messages/1")) {
        return ok(
          JSON.stringify({
            id: "1",
            labelIds: ["INBOX", "UNREAD"],
            internalDate: "1720000000000",
            payload: {
              headers: [
                { name: "From", value: "sender@x" },
                { name: "To", value: "a@x, b@x" },
                { name: "Subject", value: "Hello" },
              ],
              parts: [
                { mimeType: "text/html", body: { data: b64url("<b>hi</b>") } },
                { mimeType: "text/plain", body: { data: b64url("plain body") } },
              ],
            },
          }),
        );
      }
      // message 2: simple body, read
      return ok(
        JSON.stringify({
          id: "2",
          labelIds: ["INBOX"],
          internalDate: "1720000001000",
          payload: { headers: [{ name: "Subject", value: "Two" }], body: { data: b64url("direct") } },
        }),
      );
    });
  }

  it("provider metadata", () => {
    assert.equal(mk("t").providerId, "gmail");
    assert.equal(mk("t").isConfigured, true);
  });

  it("list-unread searches is:unread then fetches each message full", async () => {
    const http = buildGmailHttp();
    const c = new GmailEmailConnector(gmailOptions(async () => "tok"), http);
    const msgs = await c.listUnreadAsync(10);
    assert.deepEqual(msgs.map((m) => m.messageId), ["1", "2"]);
    // first message: prefers text/plain part, parses headers, UNREAD label
    assert.equal(msgs[0].from, "sender@x");
    assert.deepEqual(msgs[0].to, ["a@x", "b@x"]);
    assert.equal(msgs[0].subject, "Hello");
    assert.equal(msgs[0].bodyText, "plain body");
    assert.equal(msgs[0].unread, true);
    assert.deepEqual(msgs[0].labels, ["INBOX", "UNREAD"]);
    assert.equal(msgs[0].receivedUtc.toISOString(), new Date(1720000000000).toISOString());
    // second: direct body, read
    assert.equal(msgs[1].bodyText, "direct");
    assert.equal(msgs[1].unread, false);
    // the first outbound request was the is:unread search
    assert.ok(http.requests[0].url.includes("q=is%3Aunread"));
  });

  it("validates args and token", async () => {
    const c = new GmailEmailConnector(gmailOptions(async () => "tok"), buildGmailHttp());
    await assert.rejects(() => c.searchAsync("   ", 5), /query required/);
    await assert.rejects(() => c.searchAsync("x", 0), /max/);
    const cNull = new GmailEmailConnector(gmailOptions(async () => null), buildGmailHttp());
    await assert.rejects(() => cNull.listUnreadAsync(5), /token unavailable/);
  });

  it("mark-read POSTs removeLabelIds UNREAD", async () => {
    let body = "";
    const http = new FakeHttp((r) => {
      body = r.body ?? "";
      return ok("{}");
    });
    const c = new GmailEmailConnector(gmailOptions(async () => "tok"), http);
    await c.markReadAsync("1");
    assert.deepEqual(JSON.parse(body), { removeLabelIds: ["UNREAD"] });
    await assert.rejects(() => c.markReadAsync(""), /messageId required/);
  });
});

describe("MsGraphEmailConnector", () => {
  it("reads unread messages with from/to/labels/isRead mapping", async () => {
    const http = new FakeHttp((r) => {
      assert.ok(r.url.includes("mailFolders('Inbox')"));
      return ok(
        JSON.stringify({
          value: [
            {
              id: "g1",
              subject: "Sub",
              from: { emailAddress: { address: "from@x" } },
              toRecipients: [{ emailAddress: { address: "to@x" } }],
              receivedDateTime: "2026-07-10T12:00:00Z",
              isRead: false,
              categories: ["Work"],
              body: { content: "html body" },
            },
          ],
        }),
      );
    });
    const c = new MsGraphEmailConnector(msGraphEmailOptions(async () => "t"), http);
    const [m] = await c.listUnreadAsync(10);
    assert.equal(m.from, "from@x");
    assert.deepEqual(m.to, ["to@x"]);
    assert.equal(m.subject, "Sub");
    assert.equal(m.bodyText, "html body");
    assert.equal(m.unread, true); // isRead:false → unread
    assert.deepEqual(m.labels, ["Work"]);
    assert.equal(m.receivedUtc.toISOString(), "2026-07-10T12:00:00.000Z");
  });

  it("falls back to bodyPreview and MinValue date, and read when isRead true", async () => {
    const http = new FakeHttp(() =>
      ok(JSON.stringify({ value: [{ id: "g2", subject: "S", bodyPreview: "prev", isRead: true }] })),
    );
    const c = new MsGraphEmailConnector(msGraphEmailOptions(async () => "t"), http);
    const [m] = await c.searchAsync("x", 5);
    assert.equal(m.bodyText, "prev");
    assert.equal(m.unread, false);
    assert.equal(m.receivedUtc.getTime(), DateTimeOffsetMinValue.getTime());
  });

  it("mark-read PATCHes isRead:true", async () => {
    let method = "";
    let body = "";
    const http = new FakeHttp((r) => {
      method = r.method;
      body = r.body ?? "";
      return ok("{}");
    });
    const c = new MsGraphEmailConnector(msGraphEmailOptions(async () => "t"), http);
    await c.markReadAsync("g1");
    assert.equal(method, "PATCH");
    assert.deepEqual(JSON.parse(body), { isRead: true });
  });
});

// ── IMAP over a fake transport ────────────────────────────────────────────────

class FakeImap implements IImapTransport {
  connected = false;
  authed = false;
  disconnected = false;
  seenFlagged: number[] = [];
  lastAccess: ImapFolderAccess | null = null;
  lastQuery: ImapSearchQuery | null = null;
  constructor(private store: ImapMessageSummary[]) {}
  connectAsync(): Promise<void> {
    this.connected = true;
    return Promise.resolve();
  }
  authenticateAsync(): Promise<void> {
    this.authed = true;
    return Promise.resolve();
  }
  searchAsync(_folder: string, access: ImapFolderAccess, query: ImapSearchQuery): Promise<readonly number[]> {
    this.lastAccess = access;
    this.lastQuery = query;
    let hits = this.store;
    if (query.kind === "not-seen") hits = hits.filter((m) => ((m.flags ?? 0) & ImapMessageFlags.Seen) === 0);
    else hits = hits.filter((m) => m.subject.includes(query.text) || (m.body ?? "").includes(query.text));
    return Promise.resolve(hits.map((m) => m.uid));
  }
  fetchAsync(_folder: string, uids: readonly number[]): Promise<readonly ImapMessageSummary[]> {
    // Return in the exact order requested (already sliced/ordered by the connector).
    const byId = new Map(this.store.map((m) => [m.uid, m]));
    return Promise.resolve(uids.map((u) => byId.get(u)!).filter(Boolean));
  }
  addSeenFlagAsync(_folder: string, uid: number): Promise<void> {
    this.seenFlagged.push(uid);
    return Promise.resolve();
  }
  disconnectAsync(): Promise<void> {
    this.disconnected = true;
    return Promise.resolve();
  }
}

describe("ImapEmailConnector", () => {
  const opts = imapOptions("imap.example.com", 993, true, "alice", "pw", "INBOX");
  const store: ImapMessageSummary[] = [
    { uid: 5, from: ["s5@x"], to: ["a@x"], subject: "alpha", date: new Date("2026-07-01T00:00:00Z"), flags: ImapMessageFlags.Recent, body: "hello alpha" },
    { uid: 9, from: ["s9@x"], to: ["b@x", "c@x"], subject: "beta", date: new Date("2026-07-02T00:00:00Z"), flags: ImapMessageFlags.Seen | ImapMessageFlags.Flagged, body: "beta body" },
    { uid: 7, from: ["s7@x"], to: [], subject: "gamma alpha", date: null, flags: null, body: null },
  ];

  it("provider metadata + isConfigured", () => {
    const c = new ImapEmailConnector(opts, new FakeImap([]));
    assert.equal(c.providerId, "imap");
    assert.equal(c.isConfigured, true);
    assert.equal(new ImapEmailConnector(imapOptions("", 993, true, "", ""), new FakeImap([])).isConfigured, false);
  });

  it("list-unread searches NotSeen read-only, slices descending by uid, and maps fields", async () => {
    const t = new FakeImap(store);
    const c = new ImapEmailConnector(opts, t);
    const msgs = await c.listUnreadAsync(10);
    assert.equal(t.lastAccess, ImapFolderAccess.ReadOnly);
    assert.equal(t.lastQuery?.kind, "not-seen");
    // uids 5 and 7 are not-seen (9 is Seen). Descending: 7, 5.
    assert.deepEqual(msgs.map((m) => m.messageId), ["7", "5"]);
    // uid 7 has null flags → unread stays false (C#: Flags.HasValue && …)
    assert.equal(msgs[0].unread, false);
    assert.deepEqual(msgs[0].to, []);
    assert.equal(msgs[0].bodyText, "");
    // uid 5: Recent flag → label "Recent", not seen → unread true
    assert.equal(msgs[1].unread, true);
    assert.deepEqual(msgs[1].labels, ["Recent"]);
    assert.equal(msgs[1].from, "s5@x");
    assert.ok(t.disconnected);
  });

  it("search matches subject/body, respects max, and derives flag labels", async () => {
    const t = new FakeImap(store);
    const c = new ImapEmailConnector(opts, t);
    const msgs = await c.searchAsync("alpha", 1); // matches uid5 (subj+body) & uid7 (subj); desc→7 first, max 1
    assert.deepEqual(msgs.map((m) => m.messageId), ["7"]);
    // Verify flag decoding on the Seen|Flagged message via a direct fetch path.
    const all = await c.searchAsync("beta", 10);
    assert.deepEqual(all[0].labels, ["Seen", "Flagged"]);
    assert.equal(all[0].unread, false); // Seen set
  });

  it("mark-read requires a numeric UID and flags Seen", async () => {
    const t = new FakeImap(store);
    const c = new ImapEmailConnector(opts, t);
    await c.markReadAsync("9");
    assert.deepEqual(t.seenFlagged, [9]);
    await assert.rejects(() => c.markReadAsync("not-a-uid"), /IMAP UID/);
    await assert.rejects(() => c.markReadAsync(""), /messageId required/);
  });
});
